umh-core/pkg/fsm/agent/reconcile.go

```diff
@@ -0,0 +1,289 @@
+	if ctx.Err() != nil {
+		return ctx.Err(), false
+	}
+
+	// Step 1: If there's a lastError, see if we've waited enough.
+	if a.baseFSMInstance.ShouldSkipReconcileBecauseOfError(tick) {
+		err := a.baseFSMInstance.GetBackoffError(tick)
+		a.baseFSMInstance.GetLogger().Debugf("Skipping reconcile for agent: %v", err)
+
+		// if it is a permanent error, start the removal process and reset the error
+		if backoff.IsPermanentFailureError(err) {
+			// If already in a terminal state with permanent error, force remove
+			if a.IsRemoved() || a.IsRemoving() || a.IsMonitoringStopped() {
+				a.baseFSMInstance.GetLogger().Errorf("Agent is already in a terminal state, force removing it")
+				// In a real implementation, there might be additional cleanup here
+				return err, false
+			} else {
+				a.baseFSMInstance.GetLogger().Errorf("Agent is not in a terminal state, resetting state and removing it")
+				a.baseFSMInstance.ResetState()
+				a.Remove(ctx)
+				return nil, false // Try to reconcile towards a monitoring_stopped/removed state
+			}
+		}
+		return nil, false
+	}
+
+	// Step 2: Detect external changes.
+	if err := a.reconcileExternalChanges(ctx, tick); err != nil {
+		a.baseFSMInstance.SetError(err, tick)
+		a.baseFSMInstance.GetLogger().Errorf("Error reconciling external changes: %s", err)
+		return nil, false // Continue reconciling
+	}
+
+	// Step 3: Attempt to reconcile the state.
+	err, reconciled = a.reconcileStateTransition(ctx)
+	if err != nil {
+		// If the instance is removed, we don't want to return an error
+		if err == fsm.ErrInstanceRemoved {
+			return nil, false
+		}
+
+		a.baseFSMInstance.SetError(err, tick)
+		a.baseFSMInstance.GetLogger().Errorf("Error reconciling state: %s", err)
+		return nil, false
+	}
+
+	// Everything went well, reset any errors
+	a.baseFSMInstance.ResetState()
+
+	return nil, reconciled
+}
+
+// reconcileExternalChanges checks if the AgentInstance service status has changed
+// externally and updates the observed state.
+func (a *AgentInstance) reconcileExternalChanges(ctx context.Context, tick uint64) error {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(ComponentAgentInstance, "reconcileExternalChanges", time.Since(start))
+	}()
+
+	// Create a timeout context to prevent blocking for too long
+	observedStateCtx, cancel := context.WithTimeout(ctx, AgentUpdateObservedStateTimeout)
+	defer cancel()
+
+	// First update agent's observed state and check for config changes
+	configErr := a.updateAgentObservedState(observedStateCtx, tick)
+	if configErr != nil {
+		// Log the error for observability
+		a.baseFSMInstance.GetLogger().Errorf("Failed to update agent observed state: %v", configErr)
+
+		// If we're in active state and hit config errors, transition to degraded
+		currentState := a.baseFSMInstance.GetCurrentFSMState()
+		if currentState == OperationalStateActive {
+			a.baseFSMInstance.GetLogger().Warnf("Transitioning agent to degraded state due to configuration errors")
```

Comment:
double check this with the backoff behaviour. if you throw an error, it will result in skipping reconcilations.